### PR TITLE
Code cleanup + unit tests.

### PR DESCRIPTION
**Pull request by deepesh-kn**
Doing some cleanup, but please review.
Please provide feedback.
- Clean up of StakeHelper
- Moved stake code to Facilitator.js
- Added unit test cases for facilitator functions.
Fixes: #29 and #30

### CODE REVIEW

File: .prettierrc.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "printWidth": 120,
+  "printWidth": 79,
```
Comment (hobofan) [inline .prettierrc.json:2]: Why 79 and not 80?

File: libs/Facilitator/Facilitator.js [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+   */
+  async stake(
+    staker,
+    amount,
+    beneficiary,
+    gasPrice = '0',
+    gasLimit = '0',
+    unlockSecret,
+    facilitator,
+    gas = '7000000'
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js]: `gasPrice`,`gasLimit` and `gas` should go together at the end and optimally into a `txOptions` object

File: libs/Facilitator/Facilitator.js [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+  ) {
+    if (!web3.utils.isAddress(staker)) {
+      throw new Error('Invalid staker address.');
+    }
+
+    this.stakeAmount = new BN(amount);
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js:90]: If `this` is used in the function, it should be bound in the constructor like this:
```
this.stake = this.stake.bind(this);
```
This also gives us a cleaner way to ensure the correct scoping of `this` without the
`oThis = this;` pattern.

Same goes for all the other functions.

File: libs/Facilitator/Facilitator.js [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+   * @function stake
+   *
+   * Performs the stake process.
+   *
+   * @param {string} staker Staker address.
+   * @param {string} amount Stake amount.
+   * @param {string} beneficiary Beneficiary address for minting tokens.
+   * @param {string} gasPrice Gas price for reward calculation.
+   * @param {string} gasLimit Mas gas for reward calculation.
+   * @param {string} unlockSecret Unlock secret that will be used for progress stake.
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js]: The optional arguments should be documented as such with the jsdoc synatx:
```
 * @param {string} [unlockSecret] ...
```

File: libs/Facilitator/Facilitator.js [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+   * @param {string} gas The gas provided for the transaction execution
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js]: Documentation for `facilitator` is missing

File: libs/Facilitator/Facilitator.js [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+   *
+   * Approves gateway address for the stake amount transfer.
+   *
+   * @param {string} stakerAddress Staker account address.
+   * @param {string} stakeAmount Stake amount.
+   * @param {string} gas The gas provided for the transaction execution
+   *
+   * @returns {Promise} Promise object.
+   */
+  async approveStakeAmount(stakerAddress, stakeAmount, gas = '7000000') {
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js]: Didn't we say that this should stay in `StakeHelper` (or move into a `Staker` module)?

File: libs/Facilitator/Facilitator.js [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+          this.bounty.toString(10)
+        );
+        await this.approveStakeAmount(
+          this.stakerAddress,
+          this.stakeAmount.toString(10),
+          gas
+        );
+      }
+    } else {
+      await this.approveStakeAmount(
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js]: This seems like an odd case, and I don't think that `stake` should handle that (same for line 137). I think the optimization for when `staker` and `facilitator` are the same is fine, but I think for the case where `facilitator` != `staker`, we can also assume that `facilitator` doesn't have access to the `staker` account, and the stake amount has been approved beforehand.

File: libs/Facilitator/Facilitator.js [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+   * @function approveBountyAmount
+   *
+   * Approves gateway address for the bounty amount transfer.
+   *
+   * @param {string} facilitator Facilitator address.
+   * @param {string} gas The gas provided for the transaction execution
+   *
+   * @returns {Promise} Promise object.
+   */
+  async approveBountyAmount(facilitator, gas = '7000000') {
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js]: I think we should check if the current value of `allowance` is smaller than `bounty` before doing a `approve` transaction, as the `approve` might actually lower the `allowance`. A common optimization for workers I've seen is that you approve a much bigger pool of of tokens up front and then use that allowance up over time, which is not possible if we set it to the conrete amount every time.

Same for `approveStakeAmount`.

File: .eslintrc [hunk 9]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (schemar) [inline .eslintrc:1]: In other projects we use:
```json
{
    "extends": [
        "airbnb-base"
    ],
    "rules": {
        "indent": [
            "error",
            4
        ],
        "no-console": "off",
        "no-underscore-dangle": "off",
        "import/no-extraneous-dependencies": [
            "error",
            {
                "devDependencies": true
            }
        ]
    },
    "env": {
        "mocha": true,
        "node": true,
        "es6": true
    },
    "globals": {
        "artifacts": false,
        "contract": false,
        "assert": false
    }
}
```

I suggest to use that everywhere.

File: .eslintrc.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (schemar) [inline .eslintrc.json]: What's the purpose of this file?

File: libs/Facilitator/Facilitator.js [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+
+const BN = require('bn.js');
+const web3 = require('web3');
+const Contracts = require('../../libs/Contracts');
+const Utils = require('../utils/Utils.js');
+
+/**
+ * Class to facilitate stake and mint.
+ *
+ * @class
```
Comment (schemar) [inline libs/Facilitator/Facilitator.js]: Not sure this is required for classes. I think this is supposed to be used with constructor functions to mark them as classes.

File: libs/Facilitator/Facilitator.js [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+ */
+class Facilitator {
+  /**
+   * Constructor for facilitator.
+   * @constructor
```
Comment (schemar) [inline libs/Facilitator/Facilitator.js]: same as above

File: libs/Facilitator/Facilitator.js [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+   *
+   * @param {Object} originWeb3 Origin chain web3 object.
+   * @param {Object} auxiliaryWeb3 Auxiliary chain web3 object.
+   * @param {string} Gateway contract address.
+   * @param {Object} CoGateway contract address.
```
Comment (schemar) [inline libs/Facilitator/Facilitator.js]:
```suggestion
   * @param {string} CoGateway contract address.
```

File: libs/Facilitator/Facilitator.js [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.contracts = new Contracts(originWeb3, auxiliaryWeb3);
+    this.gatewayAddress = gatewayAddress;
+    this.coGatewayAddress = coGatewayAddress;
+  }
+
+  /**
+   * @function stake
```
Comment (schemar) [inline libs/Facilitator/Facilitator.js]: Is this required? (everywhere)

File: libs/Facilitator/Facilitator.js (hunk 2 above)
Comment (schemar) [inline libs/Facilitator/Facilitator.js]: Where does this number come from? (`7000000`)

File: libs/Facilitator/Facilitator.js [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+      throw new Error('Invalid account address.');
+    }
+
+    const gatewayContract = this.contracts.Gateway(this.gatewayAddress);
+
+    return gatewayContract.methods
+      .getNonce(accountAddress)
+      .call()
+      .then((nonce) => {
+        console.log(`\t - Gateway Nonce for ${accountAddress}:`, nonce);
```
Comment (hobofan) [inline libs/Facilitator/Facilitator.js]: Can we please get rid of all the `console.log`s? They make it impossible to write an application with a clean output.
If we want to provide the additonal insight to library users that need it, we should be using a proper logging library like loglevel or winston.

File: libs/Contracts.js [hunk 16]
```diff
@@ -12,6 +12,18 @@ class Contracts {
     oThis.auxiliaryWeb3 = Contracts._getWeb3(auxiliaryWeb3);
   }
 
+  ValueToken(address, options) {
+    const oThis = this;
```
Comment (schemar) [inline libs/Contracts.js]: should we adopt this bad habit?

File: libs/Facilitator/Facilitator.js (hunk 15 above)
Comment (schemar) [inline libs/Facilitator/Facilitator.js]: Why are we logging this? And if we want to log: why don't we use a logging lib?

**Review by schemar — COMMENTED** [review]
I can only do a high-level review, as I don't understand the details of this project. It looks good to me. I just added some general questions.

File: test/facilitator/approveBountyAmount.js [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+    // Mock facilitator.getBaseToken method to return expected base token address.
+    sinon.stub(facilitator, 'getBaseToken').callsFake(() => {
+      return baseTokenAddress;
+    });
+
+    // Mock an instance of BaseToken contract.
+    mockBaseTokenContract = sinon.mock(
+      facilitator.contracts.BaseToken(baseTokenAddress)
+    );
+    const baseTokenContract = mockBaseTokenContract.object;
```
Comment (hobofan) [inline test/facilitator/approveBountyAmount.js]: I might be missing something here, but it looks like the `mockBaseTokenContract` is never used?

File: test/facilitator/approveBountyAmount.js [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+  const tearDown = function() {
+    // Restore all mocked and spy objects.
+    mockBaseTokenContract.restore();
+    mockTx.restore();
+    spy.restore();
+  };
+
+  beforeEach(() => {
+    // runs before each test in this block
+    web3 = new Web3(new Web3.providers.HttpProvider('http://127.0.0.1:9546'));
```
Comment (hobofan) [inline test/facilitator/approveBountyAmount.js]: Is this a running instance?

File: test/facilitator/approveBountyAmount.js [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+    gatewayAddress = '0x52c50cC9bBa156C65756abd71b172B6408Dde006';
```
Comment (hobofan) [inline test/facilitator/approveBountyAmount.js]: Where do the addresses come from?

**Review by hobofan — CHANGES_REQUESTED** [review]
Tests look good 👍 Otherwise I left a few comments that might need discussing.

File: .prettierrc.json (hunk 1 above)
Comment (schemar) [reply .prettierrc.json:2]: @benjaminbollen

File: libs/Facilitator/Facilitator.js (hunk 2 above)
Comment (deepesh-kn) [reply libs/Facilitator/Facilitator.js]: Removed this.

File: libs/Facilitator/Facilitator.js (hunk 15 above)
Comment (deepesh-kn) [reply libs/Facilitator/Facilitator.js]: I removed the logs for now.

File: .prettierrc.json (hunk 1 above)
Comment (deepesh-kn) [reply .prettierrc.json:2]: As discussed, keeping it as 79 and with 2 tab spaces.

File: libs/Facilitator/Facilitator.js (hunk 2 above)
Comment (deepesh-kn) [reply libs/Facilitator/Facilitator.js]: Moved all to txOption.

File: libs/Facilitator/Facilitator.js (hunk 6 above)
Comment (deepesh-kn) [reply libs/Facilitator/Facilitator.js]: I moved all the methods to StakeHelper, The facilitator now just stitches the call.

File: test/facilitator/approveBountyAmount.js (hunk 18 above)
Comment (deepesh-kn) [reply test/facilitator/approveBountyAmount.js]: Removed this from few places.
In StakeHelper tests, when I create a mock object, it expects the provider. So it is still there in few tests.

**Comment by deepesh-kn** [comment]
PR is available for review again. 
- I have move few functions to StakeHelper. 
- Facilitator just stitches the stake by calling StakeHelper functions.
- Updated tests to check the mocked functions were called with expected params.

**Comment by schemar** [comment]
> PR is available for review again.
> 
>     * I have move few functions to StakeHelper.
> 
>     * Facilitator just stitches the stake by calling StakeHelper functions.
> 
>     * Updated tests to check the mocked functions were called with expected params.

What is `StakeHelper`? Shouldn't *that* be called facilitator?

**Comment by benjaminbollen** [comment]
> What is StakeHelper? Shouldn't that be called facilitator?

the facilitator is wider than the StakeHelper; eg, the RedeemHelper etc can all be used by facilitator too

**Comment by schemar** [comment]
> > What is StakeHelper? Shouldn't that be called facilitator?
> 
> the facilitator is wider than the StakeHelper; eg, the RedeemHelper etc can all be used by facilitator too

So it's a `Staker`?

File: libs/Facilitator/Facilitator.js (hunk 3 above)
Comment (hobofan) [reply libs/Facilitator/Facilitator.js:90]: The oThis is gone as far as I can see, but `stake` is still not bound in the `Facilitator` constructor.

**Review by hobofan — APPROVED** [review]
`stake` should be bound in `Facilitator` constructor, otherwise LGTM 🏦

File: libs/Facilitator/Facilitator.js (hunk 3 above)
Comment (deepesh-kn) [reply libs/Facilitator/Facilitator.js:90]: Added ✅

**Pull request merged**